File: api/app/schemas/career.py
from datetime import datetime

from pydantic import BaseModel, Field


class CareerRecordRead(BaseModel):
    wins: int = 0
    losses: int = 0
    ties: int = 0
    win_pct: float = 0.0


class CareerProfileRead(BaseModel):
    user_id: int
    display_name: str
    username: str | None = None
    member_since: datetime
    record: CareerRecordRead
    leagues: dict[str, int]
    drafts: dict[str, int]
    trades: dict[str, int]
    waivers: dict[str, int]
    postseason: dict[str, int]
    matchups: dict[str, int]
    scoring: dict[str, float | None]
    streaks: dict[str, int]
    rivalry: dict[str, int]


class CareerPublicProfileRead(BaseModel):
    """Authenticated-viewer-safe subset of a user's career résumé."""

    user_id: int
    display_name: str
    username: str | None = None
    member_since: datetime
    record: CareerRecordRead
    leagues: dict[str, int]
    drafts: dict[str, int]
    trades: dict[str, int]
    postseason: dict[str, int]


class CareerEventRead(BaseModel):
    id: int
    event_type: str
    title: str
    season: int | None = None
    week: int | None = None
    league_id: int | None = None
    occurred_at: datetime
    metadata: dict = Field(default_factory=dict)


class CareerEventsRead(BaseModel):
    data: list[CareerEventRead]
    total: int


class CareerLeagueRead(BaseModel):
    league_id: int
    name: str
    season: int
    status: str
    record: CareerRecordRead
    points_for: float
    final_place: int | None = None
    postseason_result: str | None = None
    rival_team_name: str | None = None
    rival_record: CareerRecordRead | None = None


class CareerLeaguesRead(BaseModel):
    data: list[CareerLeagueRead]


class CareerTrophyRead(BaseModel):
    key: str
    title: str
    season: int | None = None
    league_id: int | None = None
    subtitle: str | None = None


class CareerTrophiesRead(BaseModel):
    data: list[CareerTrophyRead]


class RivalCandidateRead(BaseModel):
    team_id: int
    team_name: str
    manager_name: str


class RivalryRead(BaseModel):
    league_id: int
    season: int
    team_id: int
    rival_team_id: int | None = None
    rival_team_name: str | None = None
    rival_manager_name: str | None = None
    selected_at: datetime | None = None
    changed_at: datetime | None = None
    can_change: bool = True
    candidates: list[RivalCandidateRead]


class RivalryUpdate(BaseModel):
    rival_team_id: int


class RivalryMatchupRead(BaseModel):
    matchup_id: int
    is_rivalry_matchup: bool
    is_championship: bool
    user_team_name: str
    rival_team_name: str | None = None
    series: CareerRecordRead
    last_meeting: dict | None = None

File: web/client/types/career.ts
export type CareerRecord = {
  wins: number;
  losses: number;
  ties: number;
  win_pct: number;
};

export type CareerProfile = {
  user_id: number;
  display_name: string;
  username?: string | null;
  member_since: string;
  record: CareerRecord;
  leagues: Record<string, number>;
  drafts: Record<string, number>;
  trades: Record<string, number>;
  waivers: Record<string, number>;
  postseason: Record<string, number>;
  matchups: Record<string, number>;
  scoring: Record<string, number | null>;
  streaks: Record<string, number>;
  rivalry: Record<string, number>;
};

export type CareerEvent = {
  id: number;
  event_type: string;
  title: string;
  season?: number | null;
  week?: number | null;
  league_id?: number | null;
  occurred_at: string;
  metadata: Record<string, unknown>;
};

export type CareerEventsResponse = { data: CareerEvent[]; total: number };

export type CareerLeague = {
  league_id: number;
  name: string;
  season: number;
  status: string;
  record: CareerRecord;
  points_for: number;
  final_place?: number | null;
  postseason_result?: string | null;
  rival_team_name?: string | null;
  rival_record?: CareerRecord | null;
};

export type CareerTrophy = {
  key: string;
  title: string;
  season?: number | null;
  league_id?: number | null;
  subtitle?: string | null;
};

export type RivalCandidate = { team_id: number; team_name: string; manager_name: string };

export type LeagueRivalry = {
  league_id: number;
  season: number;
  team_id: number;
  rival_team_id?: number | null;
  rival_team_name?: string | null;
  rival_manager_name?: string | null;
  selected_at?: string | null;
  changed_at?: string | null;
  can_change: boolean;
  candidates: RivalCandidate[];
};

export type RivalryMatchup = {
  matchup_id: number;
  is_rivalry_matchup: boolean;
  is_championship: boolean;
  user_team_name: string;
  rival_team_name?: string | null;
  series: CareerRecord;
  last_meeting?: { week: number; own_score: number; rival_score: number; result: string } | null;
};
